File: OverlaysSpark/src/main/java/app/Level.java
/**
 * Stores the parameters and settings for each level.
 */
public class Level {

    private int level;
    private int numPics;
    private int numOnes;
    private int time;
    private int clickOptimum;
    private int clickMaximum; // wie oft darf man höchstens klicken
    private int amountOfLevels = 10;
    private int[][] levelSettings = new int[amountOfLevels][]; // [welches Level][Array mit int Einstellungen]

    public Level(int level) {
        this.level = level;
        setSettings();
        printSettings();
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public int getNumPics() {
        return this.numPics;
    }

    public int getNumOnes() {
        return this.numOnes;
    }

    public boolean getDoGenerate() {
        return true;
    }

    public int getTime() {
        return this.time;
    }

    public int getAmountOfLevels() {
        return this.amountOfLevels;
    }

    private void setSettings() {
		/* create and fill levelSettings Array
		Second Array paramters:
		[0]: numPics
		[1]: numOnes
		[2]: time
		[3]: clickMaximum
		{ numPics, numOnes, time, clickMaximum }
		*/
        // LEVEL ONE
        levelSettings[0] = new int[]{3, 2, 200};
        // LEVEL TWO
        levelSettings[1] = new int[]{3, 2, 150};
        // LEVEL THREE
        levelSettings[2] = new int[]{3, 2, 120};
        // LEVEL FOUR
        levelSettings[3] = new int[]{3, 2, 100};
        // LEVEL FIVE
        levelSettings[4] = new int[]{4, 2, 100};
        // LEVEL SIX
        levelSettings[5] = new int[]{4, 2, 100};
        // LEVEL SEVEN
        levelSettings[6] = new int[]{4, 3, 100};
        // LEVEL EIGHT
        levelSettings[7] = new int[]{5, 3, 100};
        // LEVEL NINE
        levelSettings[8] = new int[]{5, 3, 100};
        // LEVEL TEN
        levelSettings[9] = new int[]{5, 4, 100};

        // get parameters of current level
		/*
		if(level < 5){
			this.doGenerate = false;
		} else{
			this.doGenerate = true;
		}
		*/
        this.numPics = levelSettings[level][0];
        this.numOnes = levelSettings[level][1];
        this.time = levelSettings[level][2];
        
        calculateClicksForScore();
    }
    
    private void calculateClicksForScore(){
    	int maximum; // anzahl an maximal erlaubten clicks des users
    	int optimum; // schnellst mögliche bzw. optimale lösung 

    	maximum = ((int) Math.pow(2, numPics) - numOnes) * numPics;
    	optimum = numPics * numOnes;
    	
    	this.clickMaximum = maximum;
    	this.clickOptimum = optimum;
    }
    
    public int getClickMaximum() {
        return this.clickMaximum;
    }
    
    public int getClickOptimum() {
        return this.clickOptimum;
    }

    private void printSettings() {
        System.out.println("######################\nLevel Index: " + level + "\nNum Pics: " + levelSettings[level][0]
                + "\nNum Ones: " + levelSettings[level][1] + "\nTime: " + levelSettings[level][2] + "\nTotal amount of Levels: " + amountOfLevels + "\n######################");
    }
}

File: src/js/Level.js
class Level {

    constructor(level) {
        this.lvl = level
        this.numPictures = undefined
        this.numOne = undefined
        this.lvlTime = undefined
        this.clickOpt = undefined
        this.clickMax = undefined
        this.generateState = undefined
        this.amountOfLvls = 10
        this.levelSettings = new Array(this.amountOfLvls)

        for (let i = 0; i < this.amountOfLevels; i++) {
            this.levelSettings[i] = new Array(3)
        }
        this.setSettings()
        this.printSettings()
    }

    set level(level) {
        this.lvl = level
    }

    set numPics(numPics) {
        this.numPictures = numPics
    }

    get numPics() {
        return this.numPictures
    }

    set numOnes(num) {
        this.numOne = num
    }

    get numOnes() {
        return this.numOne
    }

    set time(time) {
        this.lvlTime = time
    }

    get time() {
        return this.lvlTime
    }

    set amountOfLevels(amount) {
        this.amountOfLvls = amount
    }

    get amountOfLevels() {
        return this.amountOfLvls
    }

    set clickMaximum(max) {
        this.clickMax = max
    }

    get clickMaximum() {
        return this.clickMax
    }

    set clickOptimum(opt) {
        this.clickOpt = opt
    }

    get clickOptimum() {
        return this.clickOpt
    }

    get doGenerate() {
        let state = this.generateState
        console.log("doGenerate: " + state)
        return state
    }

    calculateClicksForScore() {
        let maximum
        let optimum

        maximum = (parseInt(Math.pow(2, this.numPictures)) - this.numOne) * this.numPictures
        optimum = this.numPictures * this.numOne

        this.clickMax = maximum
        this.clickOpt = optimum
    }

    setSettings() {
        // LEVEL ONE
        this.levelSettings[0] = [3, 2, 200, false]
        // LEVEL TWO
        this.levelSettings[1] = [3, 2, 150, true]
        // LEVEL THREE
        this.levelSettings[2] = [3, 2, 120, false]
        // LEVEL FOUR
        this.levelSettings[3] = [3, 2, 100, true]
        // LEVEL FIVE
        this.levelSettings[4] = [4, 2, 100, false]
        // LEVEL SIX
        this.levelSettings[5] = [4, 2, 100, true]
        // LEVEL SEVEN
        this.levelSettings[6] = [4, 3, 100, false]
        // LEVEL EIGHT
        this.levelSettings[7] = [5, 3, 100, true]
        // LEVEL NINE
        this.levelSettings[8] = [5, 3, 100, false]
        // LEVEL TEN
        this.levelSettings[9] = [5, 4, 100, true]

        console.log("Level:", this.lvl);
        this.numPics = this.levelSettings[this.lvl][0]
        this.numOne = this.levelSettings[this.lvl][1]
        this.lvlTime = this.levelSettings[this.lvl][2]
        this.generateState = this.levelSettings[this.lvl][3]

        this.calculateClicksForScore()
    }

    printSettings() {
        console.log("######################\nLevel: " + (this.lvl) + "\nNum Pics: " + this.levelSettings[this.lvl][0] +
            "\nNum Ones: " + this.levelSettings[this.lvl][1] + "\nTime: " + this.levelSettings[this.lvl][2] + "\nMax amount of clicks: " + this.clickMax +
            "\nTotal amount of Levels: " + this.amountOfLvls + "\n######################")
    }
}